apollo-router/src/test_harness/http_snapshot.rs

```diff
@@ -0,0 +1,520 @@
+                .unwrap_or(&String::from("GET")),
+            &snapshot.request.path.as_ref().unwrap_or(&String::from("/"))
+        )
+    });
+
+    std::fs::write(path, serde_json::to_string_pretty(&snapshots)?).map_err(Into::into)
+}
+
+fn load<P: AsRef<Path>>(path: P) -> Result<Vec<Snapshot>, SnapshotError> {
+    let str = std::fs::read_to_string(path)?;
+    let result: Result<Vec<Snapshot>, SnapshotError> =
+        serde_json::from_str(&str).map_err(Into::into);
+    if let Ok(snapshots) = &result {
+        info!("Loaded {} snapshots", snapshots.len());
+    }
+    result
+}
+
+impl SnapshotServer {
+    /// Spawn the server in a new task and return. Used for tests.
+    #[cfg_attr(test, allow(unreachable_pub))]
+    pub async fn spawn<P: AsRef<Path>>(
+        snapshot_path: P,
+        base_url: Uri,
+        offline: bool,
+        update: bool,
+        include_headers: Option<Vec<String>>,
+    ) -> Self {
+        Self::inner_start(
+            snapshot_path,
+            base_url,
+            true,
+            offline,
+            update,
+            include_headers,
+            None,
+        )
+        .await
+    }
+
+    /// Start the server and block. Can be used to run the server as a standalone application.
+    pub(crate) async fn start<P: AsRef<Path>>(
+        snapshot_path: P,
+        base_url: Uri,
+        offline: bool,
+        update: bool,
+        include_headers: Option<Vec<String>>,
+        listener: Option<TcpListener>,
+    ) -> Self {
+        Self::inner_start(
+            snapshot_path,
+            base_url,
+            false,
+            offline,
+            update,
+            include_headers,
+            listener,
+        )
+        .await
+    }
+
+    /// Get the URI the server is listening at
+    #[cfg_attr(not(test), allow(dead_code))]
+    #[cfg_attr(test, allow(unreachable_pub))]
+    pub fn uri(&self) -> String {
+        format!("http://{}", self.socket_address)
+    }
+
+    async fn inner_start<P: AsRef<Path>>(
+        snapshot_path: P,
+        base_url: Uri,
+        spawn: bool,
+        offline: bool,
+        update: bool,
+        include_headers: Option<Vec<String>>,
+        listener: Option<TcpListener>,
+    ) -> Self {
+        if update {
+            info!("Running in update mode ⬆️");
+        } else if offline {
+            info!("Running in offline mode ⛔️");
+        }
+
+        let snapshot_file = snapshot_path.as_ref();
+        let snapshots: Vec<Snapshot> = load(snapshot_file).unwrap_or(vec![]);
+
+        let http_service = HttpClientService::new(
+            "test",
+            rustls::ClientConfig::builder()
+                .with_safe_defaults()
+                .with_native_roots()
+                .with_no_client_auth(),
+            Client::builder().build(),
+        )
+        .expect("can create a HttpService");
+        let app = Router::new()
+            .route("/", any(root_handler))
+            .route("/*path", any(handler))
```

Comment:
After reading axum's docs on [wildcard paths](https://docs.rs/axum/latest/axum/routing/struct.Router.html#wildcards), I see why you needed separate handlers (`root_handler` and `handler`) that defer to the actual handler (`handle`).

It might be nice to add a comment here as to the limitations of axum's path matching for reference.